dev/TreeView/ViewModel.cpp

```diff
@@ -104,24 +104,145 @@ class SelectedTreeNodeVector :
 
     // Default write methods will trigger TreeView visual updates.
     // If you want to update vector content without notifying TreeViewNodes, use "core" version of the methods.
-    void AppendCore(winrt::TreeViewNode const& node)
+    void InsertAtCore(unsigned int index, winrt::TreeViewNode const& node)
     {
-        GetVectorInnerImpl()->Append(node);
+        GetVectorInnerImpl()->InsertAt(index, node);
+
+        // Keep SelectedItems and SelectedNodes in sync
+        auto selectedItems = m_viewModel->GetSelectedItems();
+        if (selectedItems.Size() != Size())
+        {
+            auto treeViewList = winrt::get_self<TreeViewList>(m_viewModel->ListControl().get());
+            if (auto item = treeViewList->ItemFromNode(node))
+            {
+                selectedItems.InsertAt(index, item);
+            }
+        }
     }
 
     void RemoveAtCore(unsigned int index)
     {
         GetVectorInnerImpl()->RemoveAt(index);
+
+        // Keep SelectedItems and SelectedNodes in sync
+        auto selectedItems = m_viewModel->GetSelectedItems();
+        if (Size() != selectedItems.Size())
+        {
+            selectedItems.RemoveAt(index);
+        }
+    }
+};
+
+#pragma endregion
+
+// Similiar to SelectedNodesVector above, we need to make decisions before the item is inserted or removed.
+// we can't use vector change events because the event already happened when event hander gets called.
+#pragma region SelectedItemsVector
+
+typedef typename VectorOptionsFromFlag<winrt::IInspectable, MakeVectorParam<VectorFlag::Observable, VectorFlag::DependencyObjectBase>()> SelectedItemsVectorOptions;
+
+class SelectedItemsVector :
+    public ReferenceTracker<
+    SelectedItemsVector,
+    reference_tracker_implements_t<typename SelectedItemsVectorOptions::VectorType>::type,
+    typename SelectedItemsVectorOptions::IterableType,
+    typename SelectedItemsVectorOptions::ObservableVectorType>,
+    public SelectedItemsVectorOptions::IVectorOwner
+{
+    Implement_Vector_Read(SelectedItemsVectorOptions)
+
+private:
+    ViewModel* m_viewModel{ nullptr };
+
+public:
+    void SetViewModel(ViewModel* viewModel)
+    {
+        m_viewModel = viewModel;
```

Comment:
What's the lifetime of this ViewModel? Why are we allowed to take a raw untracked reference?

Reply:
Updated it to use weak_ref